### PR TITLE
(Request for comments #2): eve FTP logging

### PR DESCRIPTION
**Pull request by jlucovsky**
_Request for comments only_ (_continuation of #3821_)

This PR omits documentation; documentation updates will be included in the actual PR
This changeset is for issue [2315](https://redmine.openinfosecfoundation.org/issues/2315) and implements eve logging for FTP.

I'd appreciate feedback on the log records as well as any other comments. Currently, transaction details are contained within a node with the name of the protocol -- `ftp` or `ftp-data` -- suggestions for (1) an alternate name or (2) how to represent differently are solicited.

All currently defined commands are logged.

There are 2 new event types:

1. ftp
2. ftp-data

## Sample output (event_type ftp)
```
{
  "timestamp": "2019-03-10T15:20:21.104835-0700",
  "flow_id": 546820430437462,
  "pcap_cnt": 125,
  "event_type": "ftp",
  "src_ip": "2a01:0e34:ee97:b130:8c3e:45ea:5ac6:e301",
  "src_port": 38970,
  "dest_ip": "2001:0700:0100:0118:0000:0000:0000:0047",
  "dest_port": 21,
  "proto": "TCP",
  "tx_id": 8,
  "ftp": {
    "command": "EPSV",
    "response": "Entering Extended Passive Mode (|||51093|)",
    "response_code": "229",
    "dynamic_port": 51093
  }
}
{
  "timestamp": "2019-03-10T15:20:21.104835-0700",
  "flow_id": 546820430437462,
  "pcap_cnt": 125,
  "event_type": "ftp",
  "src_ip": "2a01:0e34:ee97:b130:8c3e:45ea:5ac6:e301",
  "src_port": 38970,
  "dest_ip": "2001:0700:0100:0118:0000:0000:0000:0047",
  "dest_port": 21,
  "proto": "TCP",
  "tx_id": 9,
  "ftp": {
    "command": "RETR",
    "request": "uio-app.js",
    "response": "Opening BINARY mode data connection for uio-app.js (4807 by",
    "response_code": "150"
  }
}
{
  "timestamp": "2019-03-10T15:20:21.104835-0700",
  "flow_id": 546820430437462,
  "pcap_cnt": 125,
  "event_type": "ftp",
  "src_ip": "2a01:0e34:ee97:b130:8c3e:45ea:5ac6:e301",
  "src_port": 38970,
  "dest_ip": "2001:0700:0100:0118:0000:0000:0000:0047",
  "dest_port": 21,
  "proto": "TCP",
  "tx_id": 10,
  "ftp": {
    "command": "SIZE",
    "request": "robots.txt",
    "response": "2",
    "response_code": "213"
  }
}
```

## Sample output (event type ftp-data)
{
  "timestamp": "2019-03-10T15:20:01.500157-0700",
  "flow_id": 1951184542888733,
  "parent_id": 546820430437462,
  "pcap_cnt": 68,
  "event_type": "ftp-data",
  "src_ip": "2a01:0e34:ee97:b130:8c3e:45ea:5ac6:e301",
  "src_port": 53697,
  "dest_ip": "2001:0700:0100:0118:0000:0000:0000:0047",
  "dest_port": 51093,
  "proto": "TCP",
  "tx_id": 0,
  "ftp-data": {
    "filename": "uio-app.js",
    "command": "RETR"
  }
}
{
  "timestamp": "2019-03-10T15:20:18.000469-0700",
  "flow_id": 1383754939613831,
  "parent_id": 546820430437462,
  "pcap_cnt": 116,
  "event_type": "ftp-data",
  "src_ip": "2001:0700:0100:0118:0000:0000:0000:0047",
  "src_port": 20,
  "dest_ip": "2a01:0e34:ee97:b130:8c3e:45ea:5ac6:e301",
  "dest_port": 41813,
  "proto": "TCP",
  "tx_id": 0,
  "ftp-data": {
    "filename": "index.html",
    "command": "RETR"
  }
}
```

### CODE REVIEW

**Comment by jasonish** [comment]
Log format comments - looks good except the following:
- FTP response code. Is this always an integer? Or is it more free-form where text could actually be in it.  If its always going to be a number we should have it as a number I think. If there is a chance it could ever be a string then its fine as is.
- The event type "ftp-data" results in a JSON key of "ftp-data", but I think we've been pretty careful about not using JSON key's with a hyphen in them, but instead underscores.  So I think this is better as "ftp_data", even if the configuration file uses ftp-data.

**Comment by victorjulien** [comment]
In the example above, is this correct? I looks like it might be truncated?
```
    "response": "Opening BINARY mode data connection for uio-app.js (4807 by",
```

File: src/app-layer-ftp.c [hunk 1]
```diff
@@ -470,9 +550,31 @@ static int FTPParseRequest(Flow *f, void *ftp_state,
 
     int direction = STREAM_TOSERVER;
     while (FTPGetLine(state) >= 0) {
-        FTPParseRequestCommand(state,
-                               state->current_line, state->current_line_len);
+        const FtpCommand *cmd_descriptor;
+
+        if (!FTPParseRequestCommand(state->current_line, state->current_line_len, &cmd_descriptor)) {
```
Comment (victorjulien) [inline src/app-layer-ftp.c]: this check means that if we encounter some custom or unsupported command we won't create a tx, right?

Comment (jlucovsky) [reply src/app-layer-ftp.c]: Correct; if the command isn't parsed there won't be a transaction.

**Comment by jlucovsky** [comment]
> In the example above, is this correct? I looks like it might be truncated?
> 
> ```
>     "response": "Opening BINARY mode data connection for uio-app.js (4807 by",
> ```

Correct ---- this is due to the use of static buffers for containing the response and request lines. I'll change this to use dynamically allocated memory for holding the strings so the full lines can be captured.

**Comment by jlucovsky** [comment]
> Log format comments - looks good except the following:
> 
> * FTP response code. Is this always an integer? Or is it more free-form where text could actually be in it.  If its always going to be a number we should have it as a number I think. If there is a chance it could ever be a string then its fine as is.
> * The event type "ftp-data" results in a JSON key of "ftp-data", but I think we've been pretty careful about not using JSON key's with a hyphen in them, but instead underscores.  So I think this is better as "ftp_data", even if the configuration file uses ftp-data.

I'll investigate if the response codes are (1) always 3 digits and (2) always numeric.

I'll modify the event type and corresponding section to use `_` instead of `-`

**Comment by jlucovsky** [comment]
Continued in #3827

**Pull request closed without merging**